internal/controllers/configuration/kongadminapi_controller_envtest_test.go

```diff
@@ -0,0 +1,410 @@
+//go:build envtest
+// +build envtest
+
+package configuration
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/bombsimon/logrusr/v2"
+	"github.com/google/uuid"
+	"github.com/samber/lo"
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	corev1 "k8s.io/api/core/v1"
+	discoveryv1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	ctrl "sigs.k8s.io/controller-runtime"
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/kong/kubernetes-ingress-controller/v2/internal/adminapi"
+	"github.com/kong/kubernetes-ingress-controller/v2/test/envtest"
+)
+
+// Note: If you want apiserver output on stdout set
+// KUBEBUILDER_ATTACH_CONTROL_PLANE_OUTPUT to true when running tests.
+
+type notifier struct {
+	lock sync.RWMutex
+	t    *testing.T
+	last []adminapi.DiscoveredAdminAPI
+}
+
+func (n *notifier) Notify(adminAPIs []adminapi.DiscoveredAdminAPI) {
+	n.lock.Lock()
+	defer n.lock.Unlock()
+	n.last = adminAPIs
+}
+
+func (n *notifier) LastNotified() []adminapi.DiscoveredAdminAPI {
+	n.lock.RLock()
+	defer n.lock.RUnlock()
+	return n.last
+}
+
+// prepareEnv starts KongAdminAPIServiceReconciler with the manager in a separate
+// goroutine.
+func prepareEnv(ctx context.Context, t *testing.T, client ctrlclient.Client, cfg *rest.Config) (
```

Comment:
I'd suggest renaming it to `startKongAdminAPIServiceReconciler` (or anything mentioning this reconciler) as it's a function on a level of the package and if we add more tests in the package, `prepareEnv` can become confusing as it's not really meant for other tests than the ones that verify this particular reconciler.

Reply:
Alternative: put the definition into the `TestKongAdminAPIController`'s body.

Reply:
Agreed, renamed.